### PR TITLE
[UT] ASAN false positive with exception

### LINKED ISSUE #42349: stack-use-after-scope in UT
```
[ RUN      ] LakeMetacacheTest.test_tablet_schema_cache
  2024-03-08 05:08:59.995521, query_id=00000000-0000-0000-0000-000000000000, fragment_instance_id=00000000-0000-0000-0000-000000000000 throws exception: std::bad_variant_access, trace:
       @         0x23b390a6  __wrap___cxa_throw
      @         0x1455357d  std::__throw_bad_variant_access()
      @         0x145535aa  std::__throw_bad_variant_access()
      @         0x22f97ec4  std::get<>()
      @         0x22f96e4f  std::get<>()
      @         0x22f91914  starrocks::lake::Metacache::lookup_delvec()
      @         0x17e4e65a  starrocks::lake::LakeMetacacheTest_test_tablet_schema_cache_Test::TestBody()
      @         0x2a294a91  testing::internal::HandleExceptionsInMethodIfSupported<>()
      @         0x2a286c36  testing::Test::Run()
      @         0x2a286da5  testing::TestInfo::Run()
      @         0x2a286e95  testing::TestSuite::Run()
      @         0x2a2873e6  testing::internal::UnitTestImpl::RunAllTests()
      @         0x2a2875ee  testing::UnitTest::Run()
      @         0x145640fc  RUN_ALL_TESTS()
      @         0x1454d5a9  starrocks::init_test_env()
      @         0x1454dea9  main
      @     0x7f488e879555  __libc_start_main
      @         0x14486029  (unknown)
      @              (nil)  (unknown)
   
  
  W0308 05:08:59.995630 488679 stack_util.cpp:350] 2024-03-08 05:08:59.995521, query_id=00000000-0000-0000-0000-000000000000, fragment_instance_id=00000000-0000-0000-0000-000000000000 throws exception: std::bad_variant_access, trace:
       @         0x23b390a6  __wrap___cxa_throw
      @         0x1455357d  std::__throw_bad_variant_access()
      @         0x145535aa  std::__throw_bad_variant_access()
      @         0x22f97ec4  std::get<>()
      @         0x22f96e4f  std::get<>()
      @         0x22f91914  starrocks::lake::Metacache::lookup_delvec()
      @         0x17e4e65a  starrocks::lake::LakeMetacacheTest_test_tablet_schema_cache_Test::TestBody()
      @         0x2a294a91  testing::internal::HandleExceptionsInMethodIfSupported<>()
      @         0x2a286c36  testing::Test::Run()
      @         0x2a286da5  testing::TestInfo::Run()
      @         0x2a286e95  testing::TestSuite::Run()
      @         0x2a2873e6  testing::internal::UnitTestImpl::RunAllTests()
      @         0x2a2875ee  testing::UnitTest::Run()
      @         0x145640fc  RUN_ALL_TESTS()
      @         0x1454d5a9  starrocks::init_test_env()
      @         0x1454dea9  main
      @     0x7f488e879555  __libc_start_main
      @         0x14486029  (unknown)
      @              (nil)  (unknown)
   
  I0308 05:09:00.365482 490323 load_channel_mgr.cpp:249] Memory consumption(bytes) limit=15863717818 current=0 peak=0
  [       OK ] LakeMetacacheTest.test_tablet_schema_cache (1006 ms)
  [----------] 1 test from LakeMetacacheTest (1006 ms total)
  
  [----------] Global test environment tear-down
  [==========] 1 test from 1 test suite ran. (1006 ms total)
  [  PASSED  ] 1 test.
  =================================================================
  ==488679==ERROR: AddressSanitizer: stack-use-after-scope on address 0x7ffcc59b8f10 at pc 0x000020ecb283 bp 0x7ffcc59b8920 sp 0x7ffcc59b8918
  WRITE of size 8 at 0x7ffcc59b8f10 thread T0
      #0 0x20ecb282 in std::_Vector_base<rocksdb::DbPath, std::allocator<rocksdb::DbPath> >::_Vector_impl_data::_Vector_impl_data() (/root/starrocks/be/ut_build_ASAN/test/starrocks_test+0x20ecb282)
      #1 0x20ec8b3a in std::_Vector_base<rocksdb::DbPath, std::allocator<rocksdb::DbPath> >::_Vector_impl::_Vector_impl(std::allocator<rocksdb::DbPath> const&) (/root/starrocks/be/ut_build_ASAN/test/starrocks_test+0x20ec8b3a)
      #2 0x20ec55ef in std::_Vector_base<rocksdb::DbPath, std::allocator<rocksdb::DbPath> >::_Vector_base(unsigned long, std::allocator<rocksdb::DbPath> const&) (/root/starrocks/be/ut_build_ASAN/test/starrocks_test+0x20ec55ef)
      #3 0x20ec2d8f in std::vector<rocksdb::DbPath, std::allocator<rocksdb::DbPath> >::vector(std::vector<rocksdb::DbPath, std::allocator<rocksdb::DbPath> > const&) (/root/starrocks/be/ut_build_ASAN/test/starrocks_test+0x20ec2d8f)
      #4 0x20ec0112 in rocksdb::ColumnFamilyOptions::ColumnFamilyOptions(rocksdb::ColumnFamilyOptions const&) (/root/starrocks/be/ut_build_ASAN/test/starrocks_test+0x20ec0112)
      #5 0x254d6bfd in rocksdb::ColumnFamilyHandleImpl::~ColumnFamilyHandleImpl() (/root/starrocks/be/ut_build_ASAN/test/starrocks_test+0x254d6bfd)
      #6 0x254d7081 in rocksdb::ColumnFamilyHandleImpl::~ColumnFamilyHandleImpl() (/root/starrocks/be/ut_build_ASAN/test/starrocks_test+0x254d7081)
      #7 0x254acc2d in rocksdb::DBImpl::CloseHelper() (/root/starrocks/be/ut_build_ASAN/test/starrocks_test+0x254acc2d)
      #8 0x254ad351 in rocksdb::DBImpl::CloseImpl() (/root/starrocks/be/ut_build_ASAN/test/starrocks_test+0x254ad351)
      #9 0x254ad43a in rocksdb::DBImpl::Close() (/root/starrocks/be/ut_build_ASAN/test/starrocks_test+0x254ad43a)
      #10 0x20eb4440 in starrocks::KVStore::~KVStore() (/root/starrocks/be/ut_build_ASAN/test/starrocks_test+0x20eb4440)
      #11 0x20eb45bd in starrocks::KVStore::~KVStore() (/root/starrocks/be/ut_build_ASAN/test/starrocks_test+0x20eb45bd)
      #12 0x21ff9dad in starrocks::DataDir::~DataDir() (/root/starrocks/be/ut_build_ASAN/test/starrocks_test+0x21ff9dad)
      #13 0x21f8c944 in starrocks::StorageEngine::stop() (/root/starrocks/be/ut_build_ASAN/test/starrocks_test+0x21f8c944)
      #14 0x1454d775 in starrocks::init_test_env(int, char**) (/root/starrocks/be/ut_build_ASAN/test/starrocks_test+0x1454d775)
      #15 0x1454dea8 in main (/root/starrocks/be/ut_build_ASAN/test/starrocks_test+0x1454dea8)
      #16 0x7f488e879554 in __libc_start_main (/lib64/libc.so.6+0x22554)
      #17 0x14486028  (/root/starrocks/be/ut_build_ASAN/test/starrocks_test+0x14486028)
  
  Address 0x7ffcc59b8f10 is located in stack of thread T0
  SUMMARY: AddressSanitizer: stack-use-after-scope (/root/starrocks/be/ut_build_ASAN/test/starrocks_test+0x20ecb282) in std::_Vector_base<rocksdb::DbPath, std::allocator<rocksdb::DbPath> >::_Vector_impl_data::_Vector_impl_data()
  Shadow bytes around the buggy address:
    0x100018b2f190: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    0x100018b2f1a0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    0x100018b2f1b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    0x100018b2f1c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    0x100018b2f1d0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  =>0x100018b2f1e0: 00 00[f8]f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 00 00
    0x100018b2f1f0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    0x100018b2f200: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    0x100018b2f210: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    0x100018b2f220: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    0x100018b2f230: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  Shadow byte legend (one shadow byte represents 8 application bytes):
    Addressable:           00
    Partially addressable: 01 02 03 04 05 06 07 
    Heap left redzone:       fa
    Freed heap region:       fd
    Stack left redzone:      f1
    Stack mid redzone:       f2
    Stack right redzone:     f3
    Stack after return:      f5
    Stack use after scope:   f8
    Global redzone:          f9
    Global init order:       f6
    Poisoned by user:        f7
    Container overflow:      fc
    Array cookie:            ac
    Intra object redzone:    bb
    ASan internal:           fe
    Left alloca redzone:     ca
    Right alloca redzone:    cb
    Shadow gap:              cc
  ==488679==ABORTING
```

### PR DESCRIPTION
**Pull request by sduzh**
Fixes #42349

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [x] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5

### CODE REVIEW

**Review by trueeyu — DISMISSED** [review]

**Review by kevincai — DISMISSED** [review]

File: be/src/util/stack_util.cpp [hunk 1]
```diff
@@ -351,11 +351,14 @@ void __wrap___cxa_throw(void* thrown_exception, void* info, void (*dest)(void*))
...
     // call the real __cxa_throw():
-#ifdef __clang__
+
+#if defined(ADDRESS_SANITIZER)
+    __interceptor___cxa_throw(thrown_exception, info, dest);
+#elif defined(__clang__)
     __real___cxa_throw(thrown_exception, info, dest);
 #elif defined(__GNUC__)
-    __real___cxa_throw(thrown_exception, info, dest);
+__real___cxa_throw(thrown_exception, info, dest);
```
Comment (kevincai) [inline be/src/util/stack_util.cpp:360]: also these two condition compiling branches have no difference in implementation.

[the remark this answers is not in the document]
Comment (sduzh) [reply be/src/util/stack_util.cpp]: updated

Comment (sduzh) [reply be/src/util/stack_util.cpp]: updated

File: be/src/util/stack_util.cpp [hunk 2]
```diff
@@ -351,11 +351,12 @@ void __wrap___cxa_throw(void* thrown_exception, void* info, void (*dest)(void*))
...
-    __real___cxa_throw(thrown_exception, info, dest);
-#elif defined(__GNUC__)
+
+#if defined(ADDRESS_SANITIZER)
+    __interceptor___cxa_throw(thrown_exception, info, dest);
+#else
     __real___cxa_throw(thrown_exception, info, dest);
 #endif
-}
+} // namespace starrocks
```
Comment (kevincai) [inline be/src/util/stack_util.cpp:360]: looks like this is wrongly added.

**Review by kevincai — APPROVED** [review]

**Comment by sduzh** [comment]
@imay Could you please help merge in this PR?  The coverage report can be ignored.

**Review by imay — APPROVED** [review]

**Pull request merged**